activerecord/lib/active_record/associations/lazy_preloader.rb

```diff
@@ -32,15 +32,23 @@ class LazyPreloader
       #
       # => SELECT `books`.* FROM `books` WHERE `author_id` IN (2, 5)
       #
+      require "weakref"
+
       class Registry
+        @map = ::Hash.new
+
+        # This is needed because rails overrides #hash method for AR
         def self.store(record, instance)
-          record.instance_variable_set :@_lazy_preloader, instance
+          reference = ::WeakRef.new record
```

Comment:
I think that the `WeakRef`s need to be for the collection of records stored in the `LazyPreloader` at https://github.com/rails/rails/pull/33527/files#diff-a96765768ab0d3346409a9af7447ccf3R56. I don't think that there was anything wrong with storing the preloaded against each of the records, unless I have missed something.